neuralpp/experiments/experimental_inference/belief_propagation.py

```diff
@@ -0,0 +1,38 @@
+from typing import List, Optional
+
+from functools import reduce
+from neuralpp.inference.graphical_model.representation.factor.product_factor import (
+    Factor
+)
+from neuralpp.inference.graphical_model.variable.discrete_variable import DiscreteVariable
+from neuralpp.inference.graphical_model.variable.variable import Variable
+
+
+class BeliefPropagation:
+
+    @staticmethod
+    def mult_factors(factors: List[Factor]):
+        return reduce(lambda x, y: x * y, factors)
+
+    @staticmethod
+    def add_factors(factors: List[Factor]):
+        return reduce(lambda x, y: x.add_with_non_identity(y), factors)
+
+    @staticmethod
+    def run(query: Variable, factors: List[Factor]):
+        return BeliefPropagation.msg_from_variable(query, None, factors)
+
+    @staticmethod
+    def msg_from_factor(factor: Factor, out_var: Variable, factors: List[Factor]):
+        neighbors = [v for v in factor.variables if v != out_var]
+        if not neighbors:
+            return factor
+        assignments = DiscreteVariable.assignments_product_dicts(neighbors)
+        msg = BeliefPropagation.mult_factors([BeliefPropagation.msg_from_variable(v, factor, factors) for v in neighbors])
```

Comment:
Here you can just create a `ProductFactor` based on the list of factors to be multiplied. You want to create a list with the incoming messages and `factor` itself.